Fixed #15727 -- Added Content-Security-Policy support.

robhudson opened the pull request:
#### Trac ticket number

ticket-15727

#### Branch description
This adds CSP support for both the enforced header and report-only header.

This is for the middleware and CSP nonce features only. A follow-up commit will add view decorators.

#### Checklist
- [x] This PR targets the `main` branch.
- [x] The commit message is written in past tense, mentions the ticket number, and ends with a period.
- [x] I have checked the "Has patch" ticket flag in the Trac system (was already set from prior efforts.
- [x] I have added or updated relevant tests.
- [x] I have added or updated relevant docs, including release notes if applicable.

apollo13 reviewed (commented):
Since @carltongibson asked nicely I started a review on this. Overall it really looks good (I only looked at the implementation itself). Personally I think the docs in docs/ref/csp.txt are overly verbose in the sense that they reexplain all the csp directives when they are documented on numerous places already (specs, mdn etc). That said, I don't have strong feelings.

RealOrangeOne commented:
>  they reexplain all the csp directives when they are documented on numerous places already (specs, mdn etc). That said, I don't have strong feelings.

I agree. Web standards are constantly shifting, and documenting ourselves the possible values for the CSP itself is going to result in it being out of sync and extra burden to maintain. Linking out is almost certainly better, for maintainers and users alike.

robhudson commented:
Thank you all. There was a request to have an example of a "good" policy, maybe as a starting point, and help explain why it's good. That's where the policy in `ref/csp.txt` came from and why I'm explaining what each means and things like "Add trusted font based CDNs if necessary" for `font-src`. Over in the `ref/settings.txt` we're linking to CSP directives on MDN.

A couple paths forward I can see...
1. Make it more clear that this is an example "base" policy and drop the explanation of directives section and link to MDN with some text suggesting the policy would need changes based on the particular project and here's a link to learn more about each directive.
2. Move this section out of `ref/csp.txt` probably to the how-to as a way to quickly get started with a secure policy that likely needs additions depending on the project's needs.
3. Or both.

robhudson commented on docs/howto/csp.txt:
```diff
@@ -0,0 +1,124 @@
+
+      MIDDLEWARE = [
+          # ...
+          "django.middleware.CSP.ContentSecurityPolicyMiddleware",
```
Thanks for catching that!

robhudson commented on django/middleware/csp.py:
```diff
@@ -0,0 +1,110 @@
+    def process_response(self, request, response):
+        # In DEBUG mode, exclude CSP headers for specific status codes that
+        # trigger the debug view.
+        exempted_status_codes = {
```
Thinking through the pros/cons...

### Exempt CSP for debug pages
Pros:
* Simple and zero maintenance for future changes to these debug changes.
* No surprise breakages when developing locally.

Cons:
* While these pages should never be seen in production environments, it does happen. Having no CSP would be less secure than a minimal CSP. That being said, I think if a project is deployed with `DEBUG=True` there are bigger problems since it can expose settings, stack traces, etc.

### Custom CSP for debug pages via view decorators
Pros:
* Good example of security discipline.

Cons:
* Maintenance.
* Could possibly break, and then you're debugging the debug page.

The pros and cons are basically the opposite of each other. But, based on that I'd probably vote towards exempting CSP for debug pages even if we could override them to avoid the maintenance and any potential issues since these should only be seen during local development.

robhudson committed:
Fixed #15727 -- Added Content Security Policy (CSP) support.

This initial work adds a pair of settings to configure specific CSP
directives for enforcing or reporting policy violations, a new
`django.middleware.csp.ContentSecurityPolicyMiddleware` to apply the
appropriate headers to responses, and a context processor to support CSP
nonces in templates for safely inlining assets.

Relevant documentation has been added for the 6.0 release notes,
security overview, a new how-to page, and a dedicated reference section.

Thanks to the multiple reviewers for their precise and valuable feedback.

Co-authored-by: Natalia <124304+nessita@users.noreply.github.com>

robhudson commented:
This is now updated.

I chose to remove the full example and keep the example minimal just to show the structure of the settings. This also allowed the removal of the explanation of the settings. I added a link to the MDN directives instead. So `docs/ref/csp.txt` should be more just "reference" material on the settings, the structure, the constants, and nonces.

nessita commented on django/middleware/constants.py:
```diff
@@ -0,0 +1,40 @@
+from enum import StrEnum
```
EDIT: the more I look into the code, the more I like the idea of a dedicated, new utils `django.utils.csp`.

Basically, I think that it would be a good design pattern to have a new module `django.utils.csp` with the constants, the `LazyNonce`, and the methods `get_policy` and `build_policy` (which will turn into functions) isolated there, and then a very thin middleware importing from there.

nessita reviewed (commented):
Thank you so much @robhudson for your constant work on this PR and your enormous patience :star2: 

I've done a code review and suggested some changes. Tomorrow I'll review the docs and I'll extend the Selenium GitHub Action to add a test entry that uses a custom settings file enforcing a "decent" CSP policy. Assuming that passes, I think this will be in great shape to land.

As you know, decorators would be the next step. It's now clear to me that including them makes a lot of sense for delivering a complete feature.

nessita commented:
Other thing to consider is that when running the tests with a custom settings file such as (constants path to be adjusted):
```python
from django.conf.constants import CSP

from test_sqlite import *  # NOQA

MIDDLEWARE = [
    "django.middleware.security.SecurityMiddleware",
    "django.contrib.sessions.middleware.SessionMiddleware",
    "django.middleware.common.CommonMiddleware",
    "django.middleware.csrf.CsrfViewMiddleware",
    "django.contrib.auth.middleware.AuthenticationMiddleware",
    "django.contrib.messages.middleware.MessageMiddleware",
    "django.middleware.clickjacking.XFrameOptionsMiddleware",
    "django.middleware.csp.ContentSecurityPolicyMiddleware",
]

SECURE_CSP = {
    "default-src": [CSP.NONE],
    "img-src": [CSP.SELF],
    "script-src": [CSP.SELF],
    "style-src": [CSP.SELF],
}
```

The following failures result from a test run (`./runtests.py --settings=test_csp`). I think these tests need to ensure the CSP settings are set to the right values independently of defaults:
```
======================================================================
FAIL: test_default (middleware.test_csp.CSPGetPolicyTest.test_default)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/usr/lib/python3.13/unittest/case.py", line 58, in testPartExecutor
    yield
  File "/usr/lib/python3.13/unittest/case.py", line 651, in run
    self._callTestMethod(testMethod)
    
  File "/usr/lib/python3.13/unittest/case.py", line 606, in _callTestMethod
    if method() is not None:
    ^^^^^^^^^^^^^^^
  File "/home/nessita/fellowship/django/tests/middleware/test_csp.py", line 143, in test_default
    self.assertEqual(self.get_policy(request, response), (None, None))
    ^^^^^^^^^^^^^^^
  File "/usr/lib/python3.13/unittest/case.py", line 907, in assertEqual
    assertion_func(first, second, msg=msg)
    ^^^^^^^^^^^^^^^
  File "/usr/lib/python3.13/unittest/case.py", line 1124, in assertTupleEqual
    self.assertSequenceEqual(tuple1, tuple2, msg, seq_type=tuple)
    ^^^^^^^^^^^^^^^
  File "/usr/lib/python3.13/unittest/case.py", line 1095, in assertSequenceEqual
    self.fail(msg)
    ~~~~~~~~~^^^^^
  File "/usr/lib/python3.13/unittest/case.py", line 732, in fail
    raise self.failureException(msg)
    ^^^^^^^^^^^^^^^
AssertionError: Tuples differ: ({'default-src': [<CSP.NONE: "'none'">], '[110 chars]None) != (None, None)

First differing element 0:
{'default-src': [<CSP.NONE: "'none'">], '[103 chars]'">]}
None

+ (None, None)
- ({'default-src': [<CSP.NONE: "'none'">],
-   'img-src': [<CSP.SELF: "'self'">],
-   'script-src': [<CSP.SELF: "'self'">],
-   'style-src': [<CSP.SELF: "'self'">]},
-  None)

======================================================================
FAIL: test_default_report_only (middleware.test_csp.CSPGetPolicyTest.test_default_report_only)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/usr/lib/python3.13/unittest/case.py", line 58, in testPartExecutor
    yield
  File "/usr/lib/python3.13/unittest/case.py", line 651, in run
    self._callTestMethod(testMethod)
    
  File "/usr/lib/python3.13/unittest/case.py", line 606, in _callTestMethod
    if method() is not None:
    ^^^^^^^^^^^^^^^
  File "/home/nessita/fellowship/django/tests/middleware/test_csp.py", line 156, in test_default_report_only
    self.assertEqual(self.get_policy(request, response), (None, nonce))
    ^^^^^^^^^^^^^^^
  File "/usr/lib/python3.13/unittest/case.py", line 907, in assertEqual
    assertion_func(first, second, msg=msg)
    ^^^^^^^^^^^^^^^
  File "/usr/lib/python3.13/unittest/case.py", line 1124, in assertTupleEqual
    self.assertSequenceEqual(tuple1, tuple2, msg, seq_type=tuple)
    ^^^^^^^^^^^^^^^
  File "/usr/lib/python3.13/unittest/case.py", line 1095, in assertSequenceEqual
    self.fail(msg)
    ~~~~~~~~~^^^^^
  File "/usr/lib/python3.13/unittest/case.py", line 732, in fail
    raise self.failureException(msg)
    ^^^^^^^^^^^^^^^
AssertionError: Tuples differ: ({'default-src': [<CSP.NONE: "'none'">], '[143 chars]ng'>) != (None, <LazyNonce: 'fLqJ9pEIPMczPQJy_W5Qng'>)

First differing element 0:
{'default-src': [<CSP.NONE: "'none'">], '[103 chars]'">]}
None

- ({'default-src': [<CSP.NONE: "'none'">],
-   'img-src': [<CSP.SELF: "'self'">],
-   'script-src': [<CSP.SELF: "'self'">],
-   'style-src': [<CSP.SELF: "'self'">]},
-  <LazyNonce: 'fLqJ9pEIPMczPQJy_W5Qng'>)
+ (None, <LazyNonce: 'fLqJ9pEIPMczPQJy_W5Qng'>)
? ++++++


======================================================================
FAIL: test_csp_defaults_off (middleware.test_csp.CSPMiddlewareTest.test_csp_defaults_off)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/usr/lib/python3.13/unittest/case.py", line 58, in testPartExecutor
    yield
  File "/usr/lib/python3.13/unittest/case.py", line 651, in run
    self._callTestMethod(testMethod)
    
  File "/usr/lib/python3.13/unittest/case.py", line 606, in _callTestMethod
    if method() is not None:
    ^^^^^^^^^^^^^^^
  File "/home/nessita/fellowship/django/tests/middleware/test_csp.py", line 192, in test_csp_defaults_off
    self.assertNotIn(HEADER, response)
    ^^^^^^^^^^^^^^^
  File "/usr/lib/python3.13/unittest/case.py", line 1181, in assertNotIn
    self.fail(self._formatMessage(msg, standardMsg))
    ^^^^^^^^^^^
  File "/usr/lib/python3.13/unittest/case.py", line 732, in fail
    raise self.failureException(msg)
    ^^^^^^^^^^^^^^^
AssertionError: 'Content-Security-Policy' unexpectedly found in <HttpResponse status_code=200, "text/html; charset=utf-8">

======================================================================
FAIL: test_csp_report_only_basic (middleware.test_csp.CSPMiddlewareTest.test_csp_report_only_basic)
With SECURE_CSP_REPORT_ONLY set to a valid value, the middleware adds a
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/usr/lib/python3.13/unittest/case.py", line 58, in testPartExecutor
    yield
  File "/usr/lib/python3.13/unittest/case.py", line 651, in run
    self._callTestMethod(testMethod)
    
  File "/usr/lib/python3.13/unittest/case.py", line 606, in _callTestMethod
    if method() is not None:
    ^^^^^^^^^^^^^^^
  File "/home/nessita/fellowship/django/django/test/utils.py", line 455, in inner
    return func(*args, **kwargs)
    ^^^
  File "/home/nessita/fellowship/django/tests/middleware/test_csp.py", line 233, in test_csp_report_only_basic
    self.assertNotIn(HEADER, response)
    ^^^^^^^^^^^^^^^
  File "/usr/lib/python3.13/unittest/case.py", line 1181, in assertNotIn
    self.fail(self._formatMessage(msg, standardMsg))
    ^^^^^^^^^^^
  File "/usr/lib/python3.13/unittest/case.py", line 732, in fail
    raise self.failureException(msg)
    ^^^^^^^^^^^^^^^
AssertionError: 'Content-Security-Policy' unexpectedly found in <HttpResponse status_code=200, "text/html; charset=utf-8">
```

nessita commented:
@robhudson I've pushed all the changes mentioned in my previous review after the conversation we've had last week. Let me know what you think!

Still pending to complete my review and land:
1. There are too many "warning" notes in the docs (reader start ignoring the warning box if there are too many), I've been rewording some but there are some more that may need further analysis.
2. As per the docs, the report directive should be `report-to` and not `report-uri`, since the latter is deprecated.
3. I would love to see how the next branch takes shape, specifically the use of decorators to handle the CSP violations from the debug views. Would you have any availability to work on that?

robhudson reviewed (commented):
Thanks for all the readjustments!

nessita commented on django/contrib/admin/tests.py:
```diff
@@ -28,6 +31,11 @@ class AdminSeleniumTestCase(SeleniumTestCase, StaticLiveServerTestCase):
+    def tearDown(self):
+        # Ensure that no CSP violations were logged in the browser.
+        self.assertEqual(self.get_browser_logs(source="security"), [])
+        return super().tearDown()
```
Removing.

nessita commented:
> Thanks for all the readjustments!

Thank you, I pushed the cleanup and reverted the `report-to` recommendation back to `report-uri`. Context for others, from Rob as PM:

> Right now `report-uri` is well supported on all browsers while `report-to` is not (boo Firefox). Also, `report-to` requires the `Reporting-Endpoints` header, a separate header, to define the endpoints, which has even less support than `report-to` oddly (Firefox and Opera don't).

nessita commented on django/conf/global_settings.py:
```diff
@@ -663,6 +663,12 @@ def gettext_noop(s):
+##################
+# CSP MIDDLEWARE #
+##################
+SECURE_CSP = {}
+SECURE_CSP_REPORT_ONLY = {}
```
I settled on empty dict.

nessita committed:
Refs #15727 -- Updated AdminSeleniumTestCase to use ContentSecurityPolicyMiddleware.

Replaced the custom CSP middleware previously used in the admin's
AdminSeleniumTestCase with the official ContentSecurityPolicyMiddleware.
This change ensures alignment with Django's built-in CSP support.

Also updates the test logic to inspect browser console logs to assert
that no CSP violations are triggered during Selenium admin tests.

nessita reviewed (commented):
@robhudson I've pushed what I consider the last batch of updates:

* Docs were perused and I tried to minimized the warning/notes boxes to ensure we call for attention when truly necessary.
* I double check that each doc type (release note, how-to, topic, ref) focused on the key points relevant for that doc type. Specifically, I kept the how-to the closest possible to "recipe style" for which I moved some of the final notes into the reference docs.
* I isolated the nonce generation in the middleware, ensuring that the nonce is not accessible via the request and only via a context variable, making mandatory the use of the context processor to use nonces (as documented). My main motivation for this is to ensure that there are no surprises or unexpected outcomes when the context processor is/is not defined ("_There should be one-- and preferably only one --obvious way to do it._")
* I removed `get_policy` and simplified the middleware logic in the for loop.
* I added an important assertion for the nonce generation (asserting that the nonce is, in fact, random).

There is an outstanding question of whether the settings should be unified into a single dict. I'm a bit exhausted right now in order to think clearly about that, but it feels like we should be able to do something like this and not disrupt the code that much:
```python
SECURE_CSP = {
    "ENFORCE": {...},
    "REPORT_ONLY": {...},
}
```
I'm very keen in knowing your thoughts on this!

@RealOrangeOne @apollo13 @carltongibson @adamchainz if you feel strongly about the settings names/amount conundrum, please voice your opinion soon(ish)!

robhudson commented:
> There is an outstanding question of whether the settings should be unified into a single dict.

I considered this approach for django-csp but in the end chose not to go that route for a few reasons. Django may have different considerations, but I'll share my thoughts.

- If in the same dict, I feel there's extra checking for the keys that made the code just slightly more complex.
- I liked that 1 setting aligned with 1 HTTP header. Since they are separate headers it is a bit less cognitive load - the enforced header is one Django setting, the report-only is the other. 
- Often, in practice, the report-only is a copy of the enforced with minor changes. Having them as separate settings makes it easy to do something like `SECURE_CSP_REPORT_ONLY = copy.deepcopy(SECURE_CSP)` and then add/edit/remove directives or values as needed.

sarahboyce commented on docs/ref/csp.txt:
```diff
@@ -0,0 +1,260 @@
+Settings
+========
+
+The :class:`~django.middleware.csp.ContentSecurityPolicyMiddleware` is
+configured using the ``SECURE_CSP`` and ``SECURE_CSP_REPORT_ONLY`` settings.
+
+:setting:`SECURE_CSP`
+
+Defines the **enforced Content Security Policy**.
+
+When set, the middleware adds a ``Content-Security-Policy`` HTTP header to
+responses. This instructs the browser to actively block any content that
+violates the configured directives.
+
+.. _content-security-policy-report-only:
+
+:setting:`SECURE_CSP_REPORT_ONLY`
+
+Defines a **report-only Content Security Policy**.
+
+When set, the middleware adds a ``Content-Security-Policy-Report-Only`` HTTP
+header to responses. This allows the browser to monitor and report policy
+violations without blocking content, making it useful for refining a policy.
+
+Most browsers log CSP violations to the developer console and can optionally
+send them to a reporting endpoint. To collect these reports, the ``report-uri``
+directive must be defined (see :ref:`csp-reports` for more details).
```
Note that we have separate docs for these settings. To reduce duplication, we can use the csrf docs as a template here:
```suggestion
The following settings can be used to control Django's CSP behavior:

* :setting:`SECURE_CSP`
* :setting:`SECURE_CSP_REPORT_ONLY`
```
Then any information that is only included here could be included in the settings docs

sarahboyce commented on docs/ref/csp.txt:
```diff
@@ -0,0 +1,260 @@
+     reports can help detect regressions, unexpected changes in behavior, or
+     potential tampering in production environments.
+
+Settings structure
```
Similar here, I'm not sure if this information is covered (or should be covered) by the settings docs

nessita replied:
I usually struggle with the difficult balance between having enough details in the settings page but not "too much". Given your commentary feels like this could be moved to the settings page.

sarahboyce commented on docs/ref/csp.txt:
```diff
@@ -0,0 +1,260 @@
+<https://developer.mozilla.org/en-US/docs/Web/HTTP/Reference/Headers/Content-Security-Policy#reporting_directives>`_
+such as ``report-uri`` that specifies where violation data should be sent.
+
+Django supports configuring these directives via the relevant settings, but
```
I think this section is not clear that this is quite `SECURE_CSP_REPORT_ONLY` specific information. For example, I think the "via the relevant settings" is "via the `SECURE_CSP_REPORT_ONLY` setting"

nessita commented on docs/ref/csp.txt:
```diff
@@ -0,0 +1,260 @@
+.. _content-security-policy:
+
```
There is a general intention/desire/goal of not use `:doc:` for docs linking and always prefer `:ref:` (otherwise docs refactors are a nightmare for example). More details on https://forum.djangoproject.com/t/using-rather-than/35204.

nessita commented on docs/releases/6.0.txt:
```diff
@@ -37,6 +37,48 @@ compatible with Django 6.0.
+declaring trusted sources of content by giving browsers strict rules about
+which scripts, styles, images, or other resources can be loaded.
+
+With the new :class:`~django.middleware.csp.ContentSecurityPolicyMiddleware`
+middleware, :func:`~django.template.context_processors.csp` context processor,
+and :setting:`SECURE_CSP` and :setting:`SECURE_CSP_REPORT_ONLY` settings, CSP
+policies can be enforced or monitored directly, without relying on third-party
+libraries or custom headers.
+
+Policies are fully configurable using simple Python dictionaries and can take
+advantage of Django-provided constants for clarity and safety.
+
+When the :class:`~django.middleware.csp.ContentSecurityPolicyMiddleware` is
+enabled, Django automatically generates and injects the CSP headers, including
+support for nonces and reporting.
```
I will rework into something simpler, but since this is the "marketiny" announcement, I would still be keen to having an explicit list of the new items we are shipping.

carltongibson commented:
@robhudson — first off this is looking really good. Thanks for all the effort! 🏅

Then, since @nessita [asked](https://github.com/django/django/pull/19393#pullrequestreview-2959632151)... 

> There is an outstanding question of whether the settings should be unified into a single dict. ... if you feel strongly about the settings names/amount conundrum, please voice your opinion soon(ish)!

OK, so in the current version of Django, if I counted right, we have 10 settings that begin with the `SECURE_` prefix. Combining them/simplifying/or-something-in-that-ballpark has come up a number of times over the years, but we've yet to act on that. 

Every time it's come up @adamchainz has had the clearest take. (He's spent more time around these settings than most, I'd say.) I think what we end up doing would probably follow his view. 

For right now however, I think we should probably stick to the existing pattern — i.e. have the **two separate settings**. Any future can be done as one to all existing settings at that time (if it happens) without us having two slightly different systems in play in the meantime. (I can imagine that a single CSP setting might not match what we'd move to anyway, so we might not avoid any migration there…) 

> Often, in practice, the report-only is a copy of the enforced with minor changes. 

This seems a good point. 

I think either way will probably be fine. Taking the whole `SECURE_` namespace, I don't have a strong view on what a better API overall looks like. (If only we had a better declarative middleware story...) —  There's a lot of switches, sure, but it's quite a complex space. 

🤷

Sorry if that's not more helpful.

nessita commented:
> For right now however, I think we should probably stick to the existing pattern — i.e. have the **two separate settings**. 

Amazing! I agree, I think it makes the reasoning of the headers configuration a bit simpler.

> Sorry if that's not more helpful.

This is super helpful, thank you! :trophy:

nessita commented on docs/releases/6.0.txt:
```diff
@@ -37,6 +37,43 @@ compatible with Django 6.0.
+processor, and policies are configured using the :setting:`SECURE_CSP` and
+:setting:`SECURE_CSP_REPORT_ONLY` settings.
+
+These settings accept simple Python dictionaries and support Django-provided
```
Thank you @smithdc1! I'll remove. Just FYI, the simple refer to "non complex format" but I agree it's not necessary.

nessita reviewed (approved):
@robhudson I think we are finally there! For me this is ready to be checked in.
Will merge closer to my end of day. Thank you for all your work! :star2:

jacobtylerwalls commented:
Hi @robhudson 👋 

Curious about the wording of the note added in middleware.txt. My understanding is that any middleware that needs to access the response header should run after `ContentSecurityPolicyMiddleware` processes the response, which would entail that such a middleware should be positioned _before_. (But the note says _after_.) Is that right?

robhudson commented:
> Hi @robhudson 👋
> 
> Curious about the wording of the note added in middleware.txt. My understanding is that any middleware that needs to access the response header should run after `ContentSecurityPolicyMiddleware` processes the response, which would entail that such a middleware should be positioned _before_. (But the note says _after_.) Is that right?

Perhaps the note could be worded a bit clearer.

Seems like there's 2 parts:
1. The actual nonce (could be a `LazyNonce` until it is accessed, then it materializes into a `true` nonce)
2. The CSP header - this gets written on the response side of the middleware if the `LazyNonce` is materialized. The header replaces the nonce sentinel if `bool(nonce)` is truthy.

During the response, a middleware placed after CSP's middleware has its response run before CSP's, so if the `LazyNonce` gets materialized prior to this, then the CSP middleware includes the nonce in the header. If it were placed before, the CSP middleware builds the header first, leaving you with a materialized nonce that is never included in the header.

If the middleware wants to read the actual CSP header value, then yes, it should be before CSP's middleware which writes the header.

Maybe better wording?
```diff
-   :ref:`csp_nonce <csp-nonce>` is positioned after it, so the nonce is
-   properly included in the response header.
+   :ref:`csp_nonce <csp-nonce>` is positioned after it, so the nonce is
+   generated before this middleware builds the response header.
```

jacobtylerwalls commented:
Ah, it hadn't occurred to me that response-phase middleware might be where a nonce gets materialized.

Maybe we should even add "since nonces are lazy,"

e.g.

```diff
-   Can be placed near the bottom, but ensure any middleware that accesses
-   :ref:`csp_nonce <csp-nonce>` is positioned after it, so the nonce is
-   properly included in the response header.
+   Can be placed near the bottom, but since the :ref:`csp_nonce <csp-nonce>` is
+   lazy, ensure any middleware that accesses it is positioned after, so the
+   nonce is generated before this middleware builds the response header.
```
